File: Task mới/DataGuard/draft-01-auditability-gap/user-study/analysis/mixed_effects.R
## DataGuard user-study mixed-effects models (preferred analysis path).
##
## Run after data collection:
##   Rscript analysis/mixed_effects.R data/processed/trials.csv \
##                                    data/processed/tlx.csv \
##                                    analysis/results/
##
## Dependencies: lme4, lmerTest, emmeans, broom.mixed.

suppressPackageStartupMessages({
  library(lme4); library(lmerTest); library(emmeans); library(broom.mixed)
  library(dplyr); library(readr); library(tidyr)
})

args <- commandArgs(trailingOnly = TRUE)
trials_path <- args[1]
tlx_path    <- args[2]
out_dir     <- args[3]
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

trials <- read_csv(trials_path, show_col_types = FALSE) %>%
  mutate(condition = factor(condition, levels = c("C0","C1","C2")),
         stratum   = factor(stratum,   levels = c("S1","S2","S3")),
         pid       = factor(pid),
         app_id    = factor(app_id))

# ---- Long-format axis-level accuracy (4 rows per trial) ----
acc_long <- trials %>%
  pivot_longer(cols = starts_with("acc_"),
               names_to = "axis", values_to = "correct") %>%
  filter(!is.na(correct))

## -------------------- H1: accuracy --------------------
cat("=== H1: accuracy ~ condition + (1|pid) + (1|app_id) ===\n")
m_acc <- glmer(correct ~ condition + (1|pid) + (1|app_id),
               family = binomial(), data = acc_long,
               control = glmerControl(optimizer = "bobyqa"))
print(summary(m_acc))

em_acc <- emmeans(m_acc, ~ condition, type = "response")
write_csv(as_tibble(em_acc), file.path(out_dir, "h1_emmeans.csv"))
write_csv(as_tibble(pairs(em_acc, adjust = "holm")),
          file.path(out_dir, "h1_contrasts.csv"))

## -------------------- H4: interaction --------------------
cat("=== H4: accuracy ~ condition * stratum + (1|pid) + (1|app_id) ===\n")
m_int <- glmer(correct ~ condition * stratum + (1|pid) + (1|app_id),
               family = binomial(), data = acc_long,
               control = glmerControl(optimizer = "bobyqa"))
print(summary(m_int))
em_int <- emmeans(m_int, ~ condition | stratum, type = "response")
write_csv(as_tibble(em_int), file.path(out_dir, "h4_emmeans.csv"))
write_csv(as_tibble(pairs(em_int, adjust = "holm")),
          file.path(out_dir, "h4_contrasts.csv"))

## -------------------- H2: workload --------------------
cat("=== H2: tlx_global ~ condition + (1|pid) ===\n")
tlx <- read_csv(tlx_path, show_col_types = FALSE) %>%
  mutate(condition = factor(condition, levels = c("C0","C1","C2")),
         pid       = factor(pid))
m_tlx <- lmer(tlx_global ~ condition + (1|pid), data = tlx)
print(summary(m_tlx))
em_tlx <- emmeans(m_tlx, ~ condition)
write_csv(as_tibble(em_tlx), file.path(out_dir, "h2_emmeans.csv"))
write_csv(as_tibble(pairs(em_tlx, adjust = "holm")),
          file.path(out_dir, "h2_contrasts.csv"))

## -------------------- Time (secondary) ---------------
cat("=== Secondary: log(rt) ~ condition + (1|pid) + (1|app_id) ===\n")
trials$rt_log <- log(trials$rt_ms)
m_rt <- lmer(rt_log ~ condition + (1|pid) + (1|app_id),
             data = trials %>% filter(is.finite(rt_log)))
print(summary(m_rt))
em_rt <- emmeans(m_rt, ~ condition)
write_csv(as_tibble(em_rt), file.path(out_dir, "rt_emmeans.csv"))
write_csv(as_tibble(pairs(em_rt, adjust = "holm")),
          file.path(out_dir, "rt_contrasts.csv"))

cat("All models fitted. Results written to ", out_dir, "\n")

File: Task mới/DataGuard/draft-01-auditability-gap/user-study/analysis/mixed_effects.py
"""Python fallback for the mixed-effects analysis (statsmodels).

The preferred analysis path is mixed_effects.R (lme4). This script is here
so that reviewers without R can reproduce the headline numbers.

Models fitted:
    H1: glm logistic with participant + app fixed effects (proxy MEM)
    H2: OLS workload ~ condition + participant FE
    Reported: per-condition adjusted means + pairwise contrasts (Holm).

Run after data collection:
    python analysis/mixed_effects.py \
        --trials data/processed/trials.csv \
        --tlx    data/processed/tlx.csv \
        --out    analysis/results/
"""
from __future__ import annotations
import argparse
import json
from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm


def to_long_accuracy(trials: pd.DataFrame) -> pd.DataFrame:
    cols = [c for c in trials.columns if c.startswith("acc_")]
    long = trials.melt(id_vars=["pid","condition","app_id","stratum"],
                       value_vars=cols, var_name="axis", value_name="correct")
    long = long.dropna(subset=["correct"])
    long["correct"] = long["correct"].astype(int)
    long["condition"] = pd.Categorical(long["condition"], categories=["C0","C1","C2"])
    long["stratum"]   = pd.Categorical(long["stratum"],   categories=["S1","S2","S3"])
    return long


def fit_h1(long: pd.DataFrame) -> dict:
    model = smf.logit("correct ~ C(condition) + C(pid) + C(app_id)", data=long).fit(disp=0)
    out = {"summary": model.summary().as_text()}
    for k, v in model.params.items():
        if "condition" in k:
            out[k] = {
                "coef": float(v),
                "se": float(model.bse[k]),
                "p":  float(model.pvalues[k]),
            }
    return out


def fit_h2(tlx: pd.DataFrame) -> dict:
    tlx["condition"] = pd.Categorical(tlx["condition"], categories=["C0","C1","C2"])
    m = smf.ols("tlx_global ~ C(condition) + C(pid)", data=tlx).fit()
    return {"summary": m.summary().as_text(),
            "coefs": {k: {"coef": float(v), "se": float(m.bse[k]), "p": float(m.pvalues[k])}
                      for k, v in m.params.items() if "condition" in k}}


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("--trials", required=True)
    p.add_argument("--tlx",    required=True)
    p.add_argument("--out",    default="analysis/results/")
    a = p.parse_args()

    out_dir = Path(a.out); out_dir.mkdir(parents=True, exist_ok=True)
    trials = pd.read_csv(a.trials)
    tlx    = pd.read_csv(a.tlx)

    long = to_long_accuracy(trials)
    h1 = fit_h1(long)
    h2 = fit_h2(tlx)
    (out_dir / "h1.txt").write_text(h1["summary"])
    (out_dir / "h2.txt").write_text(h2["summary"])
    (out_dir / "h1_coefs.json").write_text(json.dumps({k:v for k,v in h1.items() if k!="summary"}, indent=2))
    (out_dir / "h2_coefs.json").write_text(json.dumps(h2["coefs"], indent=2))
    print(f"[mem] wrote results to {out_dir}")


if __name__ == "__main__":
    main()
